nodejs/src/stream_reader.rs

```diff
@@ -0,0 +1,209 @@
+
+use neon::prelude::*;
+use pravega_client::event::reader::{Event, SegmentSlice};
+use pravega_client::event::reader::{EventReader, EventReaderError};
+use pravega_client_shared::ScopedStream;
+use std::cell::RefCell;
+use std::sync::Arc;
+use tokio::runtime::Handle;
+use tokio::sync::Mutex;
+use tracing::info;
+
+impl Finalize for EventData {}
+
+///
+/// This represents an event that was read from a Pravega Segment and the offset at which the event
+/// was read from.
+///
+#[derive(new)]
+pub(crate) struct EventData {
+    offset_in_segment: i64,
+    value: Vec<u8>,
+}
+
+impl EventData {
+    ///
+    /// Return the event data as `ArrayBuffer`.
+    ///
+    pub fn js_data(mut cx: FunctionContext) -> JsResult<JsArrayBuffer> {
+        let event_data = cx.this().downcast_or_throw::<JsBox<EventData>, _>(&mut cx)?;
+        let data = event_data.value.as_slice().clone();
+
+        Ok(JsArrayBuffer::external(&mut cx, data.to_owned()))
+    }
+
+    ///
+    /// Return the event offset in the segment.
+    ///
+    pub fn js_offset(mut cx: FunctionContext) -> JsResult<JsNumber> {
+        let event_data = cx.this().downcast_or_throw::<JsBox<EventData>, _>(&mut cx)?;
+        // this may result to a f64::MAX, but I'm not able to find a f64: From<i64> impl
+        let offset = event_data.offset_in_segment as f64;
+
+        Ok(cx.number(offset))
+    }
+
+    ///
+    /// Return the string representation.
+    ///
+    pub fn js_to_str(mut cx: FunctionContext) -> JsResult<JsString> {
+        let event_data = cx.this().downcast_or_throw::<JsBox<EventData>, _>(&mut cx)?;
+        Ok(cx.string(format!(
+            "offset {:?} data :{:?}",
+            event_data.offset_in_segment, event_data.value
+        )))
+    }
+}
+
+impl Finalize for Slice {}
+
+///
+/// This represents a segment slice which can be used to read events from a Pravega segment.
+///
+#[derive(new)]
+pub(crate) struct Slice {
+    seg_slice: Option<SegmentSlice>,
+}
+
+impl Slice {
+    fn next(&mut self) -> Option<EventData> {
+        if let Some(mut slice) = self.seg_slice.take() {
+            let next_event: Option<Event> = slice.next();
+            self.seg_slice = Some(slice);
+            next_event.map(|e| EventData {
+                offset_in_segment: e.offset_in_segment,
+                value: e.value,
+            })
+        } else {
+            info!("Empty Slice");
+            None
+        }
+    }
+
+    pub fn js_next(mut cx: FunctionContext) -> JsResult<JsBox<EventData>> {
+        // Use RefCell as self.seg_slice needs to be changed.
+        // Example at https://github.com/neon-bindings/neon/blob/0.10.0/test/napi/src/js/boxed.rs
+        let slice = cx.argument::<JsBox<RefCell<Slice>>>(0)?;
+        let event_data = slice.borrow_mut().next();
+
+        match event_data {
+            Some(event_data) => Ok(cx.boxed(event_data)),
+            // TODO: better return something like undefined instead of throwing an error?
```

Comment:
What's the standard TS patten here? Throwing seems like an odd one. Does it typically have a `hasNext` like java? If so we could do that by a peek.

Reply:
I did some research on the `Iteration protocols`. Looks like as long as we implement a `next(): IteratorResult`, we should be good to go. It's up to us to decide the internal implementation (whether throwing an error or returning some custom-defined end thing).

Would you like to have an `undefined` pattern? It should looks like this:

```javascript
            const internal_event = SliceNext(slice); // <- It is hard to type internal_event as it is a external(rust) object here.
            return {
                done: internal_event ? true : false,
                value: internal_event ? Event(internal_event) : undefined,
            };
```

We could also implement a `hasNext` function, but that requires more code on rust side to determine the status and will leave nodejs and python binding different.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols
https://medium.com/bytelimes/iterables-in-javascript-custom-iterables-97a1f15852eb

Reply:
I am not worried about the API being different in different languages. As generally you are only working with one. So we should do for NodeJs what is most natural in NodeJs regardless of Python.

Reply:
I would say the current implementation is sufficient since we write the right `next()`. There is no `hasNext` pattern in the js world.